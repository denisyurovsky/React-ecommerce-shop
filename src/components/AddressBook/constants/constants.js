export const LISTBOX_PADDING = 8;
export const DOUBLE_LISTBOX_PADDING = LISTBOX_PADDING * 2;
export const LIMIT_NUMBER_ITEMS = 8;
export const ITEM_SIZE_LG = 48;
export const ITEM_SIZE_MD = 36;
