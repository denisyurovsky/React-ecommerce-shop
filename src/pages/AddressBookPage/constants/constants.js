export const MODAL_TYPE = {
  EDIT: 'edit',
  ADD: 'add',
};

export const EMPTY_ADDRESS = {
  title: '',
  name: '',
  surname: '',
  country: { name: '', id: '' },
  city: '',
  street: '',
  building: '',
  flat: '',
  phone: '',
  zip: '',
};

export const LINKS = [
  { url: '/', text: 'Home' },
  { url: '/profile', text: 'Profile' },
  {
    url: `/profile/address-book`,
    text: 'Address book',
  },
];

export const RUSSIA = 'Russian Federation';
