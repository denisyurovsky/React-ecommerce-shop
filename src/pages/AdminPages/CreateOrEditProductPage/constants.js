export const PRICE_ERROR = 'Only numbers';
export const DESCRIPTION_ERROR = 'Only latin characters';
export const NOT_FOUND_PRODUCT_ERROR =
  'The item you are looking for was not found';
