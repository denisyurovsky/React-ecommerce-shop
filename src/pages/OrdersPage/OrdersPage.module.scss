@import "../../assets/styles/colors.scss";

.page {
  width: 100%;
}

.header {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;

  @media all and (max-width: 600px) {
    align-items: flex-start;
    flex-direction: column;
  }
}

@media all and (max-width: 600px) {
  .selectForm {
    width: 100% !important;
  }
}

.cardsContainer {
  position: relative;
  width: 100%;
}

.select {
  height: 30px;
  width: 200px;

  @media all and (max-width: 600px) {
    height: 30px;
    width: unset !important;
    margin-bottom: 22px;
  }
}
