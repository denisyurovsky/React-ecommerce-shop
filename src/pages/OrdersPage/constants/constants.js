export const breadcrumbsLinks = [
  { url: '/', text: 'Home' },
  { url: '/profile', text: 'Profile' },
  { url: '/profile/orders', text: 'My orders' },
];
