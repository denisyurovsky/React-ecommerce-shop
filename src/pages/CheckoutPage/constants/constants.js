import { pathNames } from '../../../constants/pathNames';

export const LINKS = [
  { url: '/', text: 'Home' },
  {
    url: pathNames.CHECKOUT,
    text: 'Checkout',
  },
];

export const EMPTY_ORDER = {
  userId: null,
  id: null,
  products: [],
  status: null,
  addressId: null,
  deliveryAddress: {},
  deliveryPrice: null,
  totalPrice: null,
  totalDiscountPrice: null,
  totalQuantity: null,
  deliveredAt: null,
  createdAt: null,
  updatedAt: null,
};
