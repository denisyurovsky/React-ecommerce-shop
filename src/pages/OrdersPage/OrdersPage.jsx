import { FormControl, InputLabel, MenuItem, Select } from '@mui/material';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import Breadcrumbs from '../../components/Breadcrumbs/Breadcrumbs';
import { OrdersCards } from '../../components/Orders/OrdersCards/OrdersCards';
import { Title } from '../../components/Title/Title';
import Spinner from '../../components/ui-kit/Spinner/Spinner';
import { orderState, orderStatus } from '../../helpers/constants/orderStatus';
import { getOrders, selectOrders } from '../../store/orders/ordersSlice';

import { breadcrumbsLinks } from './constants/constants';

import styles from './OrdersPage.module.scss';

export const OrdersPage = () => {
  const { orders, isLoading } = useSelector(selectOrders);
  const dispatch = useDispatch();

  const [statusFilter, setStatusFilter] = useState(0);
  const [displayedOrders, setDisplayedOrders] = useState([]);

  useEffect(() => {
    dispatch(getOrders());
  }, [dispatch]);

  useEffect(() => {
    if (orders) {
      setDisplayedOrders(orders);
    }
  }, [orders]);

  const handleSelectChange = (event) => {
    setStatusFilter(event.target.value);
    setDisplayedOrders(
      orders.filter((order) => {
        if (event.target.value === 0) {
          return true;
        }

        return order.status === event.target.value;
      })
    );
  };

  return (
    <div className={styles.page}>
      <Breadcrumbs links={breadcrumbsLinks} />
      <div className={styles.header}>
        <Title>My orders</Title>
        {orders && (
          <FormControl variant="standard" className={styles.selectForm}>
            <InputLabel id="demo-simple-select-standard-label">
              Display
            </InputLabel>
            <Select
              data-testid="ordersPageSelect"
              className={styles.select}
              labelId="demo-simple-select-standard-label"
              id="demo-simple-select-standard"
              value={statusFilter}
              onChange={handleSelectChange}
            >
              <MenuItem value={0}>All</MenuItem>
              {Object.keys(orderState).map((status) => {
                const text =
                  orderState[status] === orderState.DELIVERED
                    ? orderStatus[orderState[status]].deliveryStatus
                    : orderStatus[orderState[status]].paymentStatus;

                return (
                  <MenuItem key={orderState[status]} value={orderState[status]}>
                    {text}
                  </MenuItem>
                );
              })}
            </Select>
          </FormControl>
        )}
      </div>

      <div className={styles.cardsContainer}>
        {isLoading ? (
          <Spinner width={'100%'} height={'100%'} />
        ) : (
          <OrdersCards displayedOrders={displayedOrders} />
        )}
      </div>
    </div>
  );
};
