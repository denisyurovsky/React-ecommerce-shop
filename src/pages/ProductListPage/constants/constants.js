import { pathNames } from '../../../constants/pathNames';

export const sortTypes = {
  NEW_FIRST: 'new',
  OLD_FIRST: 'old',
  EXPENSIVE_FIRST: 'expensive',
  CHEAP_FIRST: 'cheap',
};

export const pageView = {
  MODULE_VIEW: 'module',
  LIST_VIEW: 'list',
};

export const sortObj = {
  [sortTypes.NEW_FIRST]: { field: 'createdAt', order: 'desc' },
  [sortTypes.OLD_FIRST]: { field: 'createdAt', order: 'asc' },
  [sortTypes.EXPENSIVE_FIRST]: { field: 'price', order: 'desc' },
  [sortTypes.CHEAP_FIRST]: { field: 'price', order: 'asc' },
};

export const BreadcrumbsLinks = [
  { url: '/', text: 'Home' },
  { url: pathNames.PRODUCTS, text: 'Products' },
];

export const NUMBER_ITEMS_ON_PAGE = 20;

export const MOBILE_WIDTH = 540;

export const defaultParams = {
  entity: 'products',
  filters: null,
  sort: sortObj[sortTypes.NEW_FIRST],
  currentPage: 1,
  itemsPerPage: NUMBER_ITEMS_ON_PAGE,
  q: '',
};
