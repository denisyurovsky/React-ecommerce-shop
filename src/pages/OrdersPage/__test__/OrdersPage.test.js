import { fireEvent, screen, waitFor } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import React from 'react';

import { orderState } from '../../../helpers/constants/orderStatus';
import { ordersDto } from '../../../test-utils/dto/ordersDto';
import testProducts from '../../../test-utils/dto/productsDto';
import renderWith from '../../../test-utils/renderWith';
import { OrdersPage } from '../OrdersPage';

const initialUser = {
  user: {
    id: 0,
  },
};

const usersOrders = ordersDto.filter(
  (order) => order.userId === initialUser.user.id
);

let serverUser = {
  id: 0,
  isLoading: false,
  errorOccurred: false,
};

const handlersFulfilled = [
  rest.post('/confirm-order', (req, res, ctx) => {
    const { orderId } = req.body;

    const now = new Date();

    return res(
      ctx.json({
        orderStatus: orderState.DELIVERED,
        orderId,
        deliveredAt: now,
      }),
      ctx.status(200)
    );
  }),
  rest.post('/cancel-order', (req, res, ctx) => {
    const { orderId } = req.body;

    return res(
      ctx.json({ orderId, orderStatus: orderState.CANCELLED }),
      ctx.status(200)
    );
  }),
  rest.get('/orders', (req, res, ctx) => {
    const { userId } = req.body;

    const response = [];

    ordersDto.forEach((order) => {
      if (order.userId === userId) {
        response.push(order);
      }
    });

    return res(ctx.json(usersOrders), ctx.status(200));
  }),
  rest.get('/users/1', (req, res, ctx) => {
    return res(ctx.json(serverUser), ctx.status(200));
  }),
  rest.get('/users/:userId', (req, res, ctx) => {
    return res(ctx.json(serverUser), ctx.status(200));
  }),
  rest.get('/products/', (req, res, ctx) => {
    const response = [];

    if (req.url.searchParams) {
      req.url.searchParams.forEach((item) => {
        response.push(testProducts[item]);
      });
    }

    return res(ctx.json(response), ctx.status(200));
  }),
];

const serverFulfilled = setupServer(...handlersFulfilled);

describe('OrdersPage component', () => {
  describe('snapshots', () => {
    it('renders OrdersPage without data', async () => {
      const { asFragment } = renderWith(<OrdersPage />, {
        orders: {
          isLoading: false,
          errorOccurred: false,
          orders: [],
        },
        user: initialUser,
      });

      await waitFor(async () => {
        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('Rects on button clicks correct', () => {
    beforeEach(() => {
      serverFulfilled.listen();
    });

    afterEach(() => {
      serverFulfilled.close();
    });

    it('Cancels order on click and allows to filter orders', async () => {
      const firstRender = renderWith(<OrdersPage />, {
        orders: {
          isLoading: false,
          errorOccurred: false,
          orders: usersOrders,
        },
        user: initialUser,
      });

      const waitingForDeliveryMark = await firstRender.findByText(
        'Waiting for delivery'
      );
      const waitingForPaymentMark = await firstRender.findByText(
        'Waiting for payment'
      );
      const cancelledMark = await firstRender.findByText('Cancelled');

      const deliveredMark = await firstRender.findByText('Delivered');

      expect(waitingForDeliveryMark).toBeInTheDocument();
      expect(waitingForPaymentMark).toBeInTheDocument();
      expect(cancelledMark).toBeInTheDocument();
      expect(deliveredMark).toBeInTheDocument();

      const expandButtons = await screen.findAllByTestId('ExpandMoreIcon');

      expect(expandButtons.length).toBe(4);

      fireEvent.click(expandButtons[0]);

      const cancelButtons = await screen.findAllByText('Cancel order');

      fireEvent.click(cancelButtons[0]);

      let newCancelledButton = screen.queryByText('Cancel order');

      await waitFor(() => {
        expect(newCancelledButton).toEqual(null);
      });

      let newCancelledMarks = await screen.findAllByText('Cancelled');

      await waitFor(() => {
        expect(newCancelledMarks.length).toEqual(2);
      });

      fireEvent.click(expandButtons[0]);

      const filterButton = await screen.findByText('All');

      fireEvent.mouseDown(filterButton);

      newCancelledMarks = await screen.findAllByText('Cancelled');
      await waitFor(async () => {
        expect(newCancelledMarks.length).toBe(3);
      });

      fireEvent.click(newCancelledMarks[2]);
      const newExpandButtons = await screen.findAllByTestId('ExpandMoreIcon');

      await waitFor(async () => {
        expect(newExpandButtons.length).toEqual(2);
      });

      fireEvent.mouseDown(filterButton);
      const selectAllOption = await screen.findByText('All');

      await waitFor(async () => {
        expect(selectAllOption).toBeInTheDocument();
      });

      fireEvent.click(selectAllOption);
      const allExpandButtons = await screen.findAllByTestId('ExpandMoreIcon');

      await waitFor(async () => {
        expect(allExpandButtons.length).toEqual(4);
      });
    }, 10000);
  });
});
