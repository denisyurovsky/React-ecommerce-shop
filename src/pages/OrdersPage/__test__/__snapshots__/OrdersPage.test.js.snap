// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersPage component snapshots renders OrdersPage without data 1`] = `
<DocumentFragment>
  <div
    class="page"
  >
    <div
      class="breadcrumbs"
      role="presentation"
    >
      <nav
        aria-label="breadcrumb"
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
      >
        <ol
          class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
        >
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="breadcrumb"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="breadcrumb"
              href="/profile"
            >
              Profile
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="breadcrumb"
              href="/profile/orders"
            >
              My orders
            </a>
          </li>
        </ol>
      </nav>
    </div>
    <div
      class="header"
    >
      <h2
        class="header"
        data-testid="title"
      >
        My orders
      </h2>
      <div
        class="MuiFormControl-root selectForm css-1nrlq1o-MuiFormControl-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="demo-simple-select-standard-label"
        >
          Display
        </label>
        <div
          class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl select css-1480iag-MuiInputBase-root-MuiInput-root"
          data-testid="ordersPageSelect"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="demo-simple-select-standard-label demo-simple-select-standard"
            class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
            id="demo-simple-select-standard"
            role="button"
            tabindex="0"
          >
            All
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="0"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="cardsContainer"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg container css-1tz8lv8-MuiContainer-root"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
  <div
    class="Toastify"
  />
</DocumentFragment>
`;
